crates/bevy_scene/src/scene_filter.rs

```diff
@@ -0,0 +1,278 @@
+use bevy_utils::hashbrown::hash_set::IntoIter;
+use bevy_utils::HashSet;
+use std::any::{Any, TypeId};
+
+/// A filter used to control which types can be added to a [`DynamicScene`].
+///
+/// This scene filter _can_ be used more generically to represent a filter for any given type;
+/// however, note that its intended usage with `DynamicScene` only considers [components] and [resources].
+/// Adding types that are not a component or resource will have no effect when used with `DynamicScene`.
+///
+/// [`DynamicScene`]: crate::DynamicScene
+/// [components]: bevy_ecs::prelude::Component
+/// [resources]: bevy_ecs::prelude::Resource
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub enum SceneFilter {
+    /// Represents an unset filter.
+    ///
+    /// This is the equivalent of an empty [`Denylist`] or an [`Allowlist`] containing every type—
+    /// essentially, all types are permissible.
+    ///
+    /// [Allowing] a type will convert this filter to an `Allowlist`.
+    /// Similarly, [denying] a type will convert this filter to a `Denylist`.
+    ///
+    /// [`Denylist`]: SceneFilter::Denylist
+    /// [`Allowlist`]: SceneFilter::Allowlist
+    /// [Allowing]: SceneFilter::allow
+    /// [denying]: SceneFilter::deny
+    #[default]
+    Unset,
+    /// Contains the set of permitted types by their [`TypeId`].
+    ///
+    /// Types not contained within this set should not be allowed to be saved to an associated [`DynamicScene`].
+    ///
+    /// [`DynamicScene`]: crate::DynamicScene
+    Allowlist(HashSet<TypeId>),
```

Comment:
Why should this be `TypeId` rather than `ComponentId`? The scene filter tool would be useful with dynamic components, and the component/resource types need to be registered already.

I can live with this being moved to follow-up work though.

Reply:
I believe my reasoning was purely an ergonomics one. I wanted this kind of behavior:

```rust
filter
  .allow::<Foo>()
  .allow::<Bar>()
  .allow::<Baz>();
```

And I'm not sure if it would be possible to achieve that by relying on `ComponentId` since it requires access to the world (please correct me if I'm wrong).

Reply:
Ah hmm. Right, that's much trickier.

I think in a follow-up then we should add two more enum variants that store `ComponentId` for the more advanced cases at the expense of ergonomics.